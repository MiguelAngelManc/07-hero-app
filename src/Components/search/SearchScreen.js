import React, { useMemo } from 'react'
import { useLocation } from 'react-router-dom/cjs/react-router-dom.min'
import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard'
import queryString from "query-string";
import { getHeroByName } from '../../selectors/getHeroByName'

export const SearchScreen = ({ history }) => {

    const location = useLocation()

    const { q = '' } = queryString.parse(location.search)


    const [{ search }, handleInputChange, reset] = useForm({ search: q })

    const filtered = useMemo(() => getHeroByName(q), [q])

    const handleSubmit = (e) => {
        e.preventDefault();
        history.push(`?q=${search}`)
    }

    return (
        <>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">

                <div className="col-5">

                    <h4>Search form</h4>
                    <hr />

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            placeholder="Search"
                            className="form-control"
                            name="search"
                            autoComplete="off"
                            value={search}
                            onChange={handleInputChange}
                        />
                        <button
                            type="submit"
                            className="btn mt-1 btn-block btn-outline-primary"
                        >
                            Search
                        </button>
                    </form>

                </div>


                <div className="col-7">
                    <h4>Results</h4>
                    <hr />


                    {
                        q === '' && <div className="alert alert-info"> Search a hero...</div>
                    }

                    {
                        (q !== '' && filtered.length == 0) && <div className="alert alert-danger"> No results found</div>
                    }

                    <div>
                        {filtered.map(hero => <HeroCard key={hero.id} {...hero} />)}
                    </div>






                </div>

            </div>
        </>
    )
}
