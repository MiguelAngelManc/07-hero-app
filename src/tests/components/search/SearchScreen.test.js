import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router-dom"
import { SearchScreen } from "../../../Components/search/SearchScreen"

describe('Pruebas en SearchScreen', () => {
    
    const mockHistory = {
        push:jest.fn()
    }

    test('Prueba con snapshot', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={(props)=><SearchScreen  history={mockHistory}/>}/>
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert').text().trim()).toBe('Search a hero...')
    })
    

    test('Debe de mostrar a batman y input con query string', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route path="/search" component={(props)=><SearchScreen  history={mockHistory}/>}/>
            </MemoryRouter>

        )
        
        expect(wrapper.find('input').prop('value')).toBe('batman')
        expect(wrapper).toMatchSnapshot();

    })

    test('Debe mostrar error si no encuentra al heroe', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=sdfsdf"]}>
                <Route path="/search" component={(props)=><SearchScreen  history={mockHistory}/>}/>
            </MemoryRouter>


        )
            
        expect(wrapper.find('.alert-danger').exists()).toBe(true)
        expect(wrapper).toMatchSnapshot();
    })
    
    test('Hace el push', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=sdfsdf"]}>
                <Route path="/search" component={(props)=><SearchScreen  history={mockHistory}/>}/>
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', {
            target:{
                name:'search',
                value:'batman'
            }
        })

        wrapper.find('form').simulate('submit',{preventDefault(){}})

        expect(mockHistory.push).toHaveBeenCalledWith(`?q=batman`)

    })
    

})
