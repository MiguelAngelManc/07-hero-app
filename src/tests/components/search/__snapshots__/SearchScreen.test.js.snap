// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen Debe de mostrar a batman y input con query string 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 550,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="column col-md-4"
            >
              <img
                alt="Heroe"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
              <div
                className="col-md-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Detective Comics #27
                    </small>
                  </p>
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    Mas...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en SearchScreen Debe mostrar error si no encuentra al heroe 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value="sdfsdf"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
         No results found
      </div>
      <div />
    </div>
  </div>,
]
`;

exports[`Pruebas en SearchScreen Prueba con snapshot 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
         Search a hero...
      </div>
      <div />
    </div>
  </div>,
]
`;
